src/Core/Managed/Shared/Extensibility/MetricAggregatorManager.cs

```diff
@@ -0,0 +1,295 @@
+        /// Generates telemetry object based on the metric aggregator.
+        /// </summary>
+        /// <param name="aggregator">Metric aggregator.</param>
+        /// <param name="aggregatorStats">Metric aggregator statistics calculated for a period of time.</param>
+        /// <returns>Metric telemetry object resulting from aggregation.</returns>
+        private static AggregatedMetricTelemetry CreateAggergatedMetricTelemetry(MetricAggregator aggregator, SimpleMetricStatisticsAggregator aggregatorStats)
+        {
+            if ((aggregator == null) || (aggregatorStats == null) || (aggregatorStats.Count <= 0))
+            {
+                return null;
+            }
+
+            var telemetry = new AggregatedMetricTelemetry(
+                aggregator.MetricName,
+                aggregatorStats.Count,
+                aggregatorStats.Sum,
+                aggregatorStats.Min,
+                aggregatorStats.Max,
+                aggregatorStats.StandardDeviation);
+
+            if (aggregator.Dimensions != null)
+            {
+                foreach (KeyValuePair<string, string> property in aggregator.Dimensions)
+                {
+                    telemetry.Properties.Add(property);
+                }
+            }
+
+            return telemetry;
+        }
+
+        private void Flush(bool disposing)
+        {
+            if ((!disposing) && this.disposed)
+            {
+                throw new ObjectDisposedException(nameof(MetricAggregatorManager));
+            }
+
+            try
+            {
+                this.Snapshot();
+                this.telemetryClient.Flush();
+            }
+            catch (Exception ex)
+            {
+                CoreEventSource.Log.FailedToFlushMetricAggregators(ex.ToString());
+            }
+        }
+
+        /// <summary>
+        /// Takes a snapshot of aggregators collected by this instance of the manager
+        /// and schedules the next snapshot.
+        /// </summary>
+        private Task SnapshotAndReschedule()
+        {
+            return Task.Factory.StartNew(
+                () =>
+                {
+                    try
+                    {
+                        this.Snapshot();
+                    }
+                    catch (Exception ex)
+                    {
+                        CoreEventSource.Log.FailedToSnapshotMetricAggregators(ex.ToString());
+                    }
+                    finally
+                    {
+                        this.snapshotTask.Delay = GetWaitTime();
+                        this.snapshotTask.Start(this.SnapshotAndReschedule);
+                    }
+                });
+        }
+
+        /// <summary>
+        /// Takes snapshot of all active metric aggregators and turns results into metric telemetry.
+        /// </summary>
+        private void Snapshot()
+        {
+            ConcurrentDictionary<MetricAggregator, SimpleMetricStatisticsAggregator> aggregatorSnapshot =
```

Comment:
Convert explicit type to `var` since the type name already appears in the same statement?